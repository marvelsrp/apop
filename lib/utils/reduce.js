import isArray from 'lodash/isArray';
import isPlainObject from 'lodash/isPlainObject';
import parseArray from 'rmk/utils/parseArray';
import parseObject from 'rmk/utils/parseObject';

/**
 * Parse object or array
 * @memberof utils
 * @param {Object|Array} origin Origin object or array
 * @param {Array} actions Array of transform actions
 *
 * @returns {Object|Array} Transformed object or array by actions
 */
export default (origin, actions) => {
  if (isArray(origin)) {
    return parseArray(origin, actions);
  } else if (isPlainObject(origin)) {
    return parseObject(origin, actions);
  } else {
    return origin;
  }
};