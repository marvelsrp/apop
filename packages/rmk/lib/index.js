const reduceObject = (origin, remakers) => {
  return remakers.reduce((state, remake) => {
    return remake(state);
  }, origin);
};

const reduceArray = (origin, remakers) => {
  return origin.reduce((state, element) => {
    state.push(reduceObject(element, remakers));
    return state;
  }, []);
};

const reduce = (origin, reamakers) => {
  if (isArray(origin)) {
    return reduceArray(origin, mutators);
  } else if (isObject(origin)) {
    return reduceObject(origin, reamakers);
  } else {
    throw new Error('Wrong input arguments');
  }
};

const rmk = (...reamakers) => {
  return origin => reduce(origin, reamakers);
};

export default rmk;
