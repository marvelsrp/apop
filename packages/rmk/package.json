{
  "name": "@rmk/rmk",
  "version": "1.0.0",
  "scripts": {
    "prepublish": "babel -d ./ ./lib",
    "dev": "babel -d ./ ./lib --watch",
    "test": "jest --coverage",
    "test-dev": "jest --watch"
  },
  "repository": {
    "type": "git",
    "url": "git+https://github.com/tuchk4/rmk.git"
  },
  "dependencies": {
    "@rmk/clear": "=1.0.0"
  },
  "license": "ISC",
  "main": "index.js"
}
