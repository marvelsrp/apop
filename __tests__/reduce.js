import rmk from '../lib';
import reduce from '../lib/utils/reduce';

const before = {
  id: 1,
  first_name: 'Stephen',
  last_name: 'Nelson',
  email: 'snelson0@yellowbook.com',
  gender: 'Male',
  ip_address: '74.191.108.58',
  city: 'Kapotnya',
  phrase: 'Universal 6th generation portal',
  null: null,
  undefined,
  date: new Date(Date.UTC(2001, 1, 1)),
  0: 0,
  1: 1,
  '-1': -1,
  emptyStr: '',
};

test('Object argument', () => {
  const result = reduce(before, [rmk.clear()]);
  expect(result).toMatchSnapshot();
});

test('Array argument', () => {
  const result = reduce([before], [rmk.clear()]);
  expect(result).toMatchSnapshot();
});
