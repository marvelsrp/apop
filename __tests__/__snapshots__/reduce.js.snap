// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Array argument 1`] = `
Array [
  Object {
    "-1": -1,
    "0": 0,
    "1": 1,
    "city": "Kapotnya",
    "date": 2001-02-01T00:00:00.000Z,
    "email": "snelson0@yellowbook.com",
    "first_name": "Stephen",
    "gender": "Male",
    "id": 1,
    "ip_address": "74.191.108.58",
    "last_name": "Nelson",
    "phrase": "Universal 6th generation portal",
  },
]
`;

exports[`Object argument 1`] = `
Object {
  "-1": -1,
  "0": 0,
  "1": 1,
  "city": "Kapotnya",
  "date": 2001-02-01T00:00:00.000Z,
  "email": "snelson0@yellowbook.com",
  "first_name": "Stephen",
  "gender": "Male",
  "id": 1,
  "ip_address": "74.191.108.58",
  "last_name": "Nelson",
  "phrase": "Universal 6th generation portal",
}
`;
